wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+## Client Security
+
+As noted in https://github.com/polymorpher/one-wallet/issues/5, in ONE Wallet v0.1 the authenticator cannot provide security protection if the client is compromised. A compromised client means the information stored on the client is leaked to a malicious third-party. In the current setup, the third-party may use these leaked information to easily reverse engineer (by brute force) the proof needed to perform an unauthorized transfer from ONE Wallet. The 6-digit code input from Google Authenticator would not be required.
```

Comment:
Probably the most important question that I forgot to ask before is whether is this attacker model even realistic? 

**Intuition:** If the attacker (e.g., malware) has access to the client, she will very likely be able to tamper with the client, and thus capture OTP from the user, while no brute-forcing is required. So please think about use cases in which this does not occur, while only leakage of client's data occur.

Reply:
Let's consider two attack patterns: (1) general large scale malware attack, and (2) an attack targeting a specific user for a specific purpose. It is possible that (1) may lead to (2) if the malware is undetected for a long period of time and the user, presumably a high value target, fails to act against the malware within this period of time. This scenario will be discussed separately. 

Furthermore, within each context, it is important to distinguish various levels of "compromised client" in real-world settings: (a) temporary control by adversaries for a brief window of time, (b) prolonged, limited read access by adversaries (mouse, keyboard input, and storage), (c) prolonged, unlimited, protected read access by adversaries (display is accessible by attacker, but sandboxed browser/app sessions are still protected),  (d) prolonged, unlimited, unprotected read access by adversaries. Within each of (a)-(d), it is also important to distinguish several levels of write access: (i) ability to obstruct client's network, (ii) ability to obstruct client's operation, (iii) ability to write to client's storage, (iv) ability to alter client's memory, (v) ability to alter client's code.

General operations under the wallet with Client Security should be secure against attackers under (a)(b)(c), even with write access (i)(ii)(iii)(iv). General operations do not include those exposing OTP Seed at the client (Create, Restore). For seed-related operations, the list is reduced to (a)(b) and (i)(ii)(iii), though some execution environments (including browsers) may protect a region in the display so (c) can also be included. 

Using modern browsers (user coverage >95%) and mobile phones, the situation of (d) and (iv), (v) are very unlikely, even if the malware gains a high level access in the user's system. The browsers (in conjunction with OS) and phone's operating systems do an excellent job in protecting memory access and the executable code. An attacker with root access would have better luck providing a "fake browser" shim and take control from there. That attack is easy to detect and browsers have various safeguards against that. Moreover, most devices made after 2016 have  secure enclave capability to provide a trusted execution environment. The trusted execution environment would make the execution safe (including memory access) against even attackers with root access. The support from OS is already implemented, and is available in at least macOS. The support from browser might take another 1-3 years to become generally available.

Thus, unintended leakage of stored credentials and keys to access them are the primary targets for attack, because there is simply no way the browser can protect from other programs intercepting keyboard inputs and reading from disk storage. For example, MetaMask stores a password-protected private key at a disk location: https://ethereum.stackexchange.com/questions/52658/where-does-metamask-store-the-wallet-seed-file-path#64038. A malware only needs to log the user's keyboard input and send a copy of the file at the disk location to the attacker to gain full access to the user's wallet.  A simple Google search would yield many detailed stories of how people lost their wallets due to this form of attack.

The analysis brings us back to the attack patterns in (1) and (2). The attack patterns in (1) generally rely on spotting vulnerable clients, intercepting necessary credentials, reporting credentials back to the attacker, similar to the MetaMask example above. It would be ineffective against our wallet for the reasons already explained in Client Security document. One could argue that the attack pattern can be modified and automated, such that the malware automatically intercepts the OTP(s), blocks the user's executions, and send funds to an alternative address before the user's normal execution is completed. This attack would inevitably require (d) (or (b)/(c) in some cases, discussed below) and at least (i) or (ii). On mobile, (d) is well protected by the OS and is extremely hard to achieve, as explained above. On desktop, (d) is less protected than mobile but the browser still offers good protection. In the case of (b) or (c), the attacker would have to identify and extract the OTPs in a sequence of keyboard inputs with timestamps. With the extra information provided in (c) and assumes the display region is not protected, the job becomes substantially easier. However, it can be disrupted by generating synthetic keyboard input noises and/or requiring the input to be made by mouse clicks on a protected, synthetic keypad with numbers placed at scrambled positions. Since the attacker only has limited time (60 seconds by default) to complete the attack (and very likely only one chance, since the longer it takes, the easier it is to detect), an automated program is very unlikely to succeed. The reason for time limit is as the following: in both (i) and (ii) the attacker could choose to disrupt only ONE Wallet related network connection to make the attack harder to detect. However, the attacker would have to preclude the user's client from committing the transaction altogether, as otherwise the attacker's commit-reveal would not succeed by the design of the mechanism (side note: we should require nonce==expectedNonce in the smart contract, instead of >=). This prevention of commit would at least raise suspicion. Moreover, the attacker has to complete his commit and reveal within a delay factor of REVEAL_MAX_DELAY, which is set to 60 seconds as of now. Assuming the attack succeeds, it would also be immediately noticed by the user.

An attacker following attack pattern (2) would follow a similar pattern described in the last paragraph, but exercise a greater control over whether, when, and how the attack is made and how user's legitimate transaction is blocked. It would also have the luxury to observe the keyboard and mouse input patterns (and potentially visual output from the display) and make far more intelligent and effective attempts than an automated program for the mass. Such attack has to be highly customized and carefully made. Nonetheless, it would not be able to prevent the user from noticing the attack and the attacker would not have a second chance after it is detected.  These rare, high-value, vulnerable individuals should consider utilizing Yubikey as proposed in #5 .